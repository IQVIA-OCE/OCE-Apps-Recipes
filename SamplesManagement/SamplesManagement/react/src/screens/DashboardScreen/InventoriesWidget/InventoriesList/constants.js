export const inventoriesColors = {
  AdHocInventory: '#b781d3',
  AuditedInventory: '#4ad463',
  InitialInventory: '#769ed9',
  PeriodicInventory: '#ef6e64',
}
