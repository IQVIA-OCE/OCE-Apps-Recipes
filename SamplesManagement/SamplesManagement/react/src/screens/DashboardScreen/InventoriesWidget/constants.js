export const inventoriesColors = {
  AdHocInventory: '#b781d3',
  AuditedInventory: '#4ad463',
  InitialInventory: '#769ed9',
  PeriodicInventory: '#ef6e64',
}

export const INVENTORY_TYPE = {
  InitialInventory: 'Initial Inventory',
  AdHocInventory: 'Ad Hoc Inventory',
  AuditedInventory: 'Audited Inventory',
  PeriodicInventory: 'Periodic Inventory',
};