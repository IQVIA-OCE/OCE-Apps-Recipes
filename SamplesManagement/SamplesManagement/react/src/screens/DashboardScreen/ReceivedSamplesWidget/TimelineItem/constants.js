export const recordTypes = {
  AcknowledgementOfShipment: {
    icon: 'format-list-checks',
    color: '#ef6e64',
  },
  TransferIn: {
    icon: 'format-indent-increase',
    color: '#f88962',
  },
};
