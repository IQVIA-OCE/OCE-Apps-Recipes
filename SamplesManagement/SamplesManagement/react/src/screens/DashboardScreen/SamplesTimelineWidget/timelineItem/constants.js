export const recordTypes = {
  AcknowledgementOfShipment: {
    icon: 'format-list-checks',
    color: '#ef6e64',
  },
  Adjustment: {
    icon: 'tune-vertical',
    color: '#47cfd2',
  },
  Disbursement: {
    icon: 'file-document-box-outline',
    color: '#769ed9',
  },
  TransferIn: {
    icon: 'format-indent-increase',
    color: '#f88962',
  },
  TransferOut: {
    icon: 'format-indent-decrease',
    color: '#f88962',
  },
  Order: {
    icon: 'clipboard-check',
    color: '#69cd37',
  },
  Return: {
    icon: 'package-variant-closed',
    color: '#b781d3',
  },
};
