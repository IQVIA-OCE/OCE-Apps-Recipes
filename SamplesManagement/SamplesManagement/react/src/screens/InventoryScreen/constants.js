export const DATE_TIME_FORMAT = 'MMM D, YYYY h:mm A';
export const DATE_TIME_FORMAT_WITH_SECONDS = 'MMM D, YYYY h:mm:ss A';
export const DATE_FORMAT = 'MMM D, YYYY';
export const INVENTORY_STATUS = {
  inProgress: 'In Progress',
  saved: 'Saved',
  submitted: 'Submitted',
};

export const BANNER_CONFIG = {
  inProgress: {
    variant: 'success',
    message: 'In Progress. Saved Successfully',
    visible: true,
    icon: 'checkbox-marked-circle',
  },
  saved: {
    variant: 'warning',
    message: 'Saved. Need reason for discrepancies',
    visible: true,
    icon: 'alert',
  },
  submitted: {
    variant: 'success',
    message: 'Submitted Successfully',
    visible: true,
    icon: 'checkbox-marked-circle',
  },
};

export const REASON_LIST = [
  { label: 'Theft/Loss', id: 'Theft/Loss' },
  { label: 'Damaged', id: 'Damaged' },
  { label: 'Disbursement Error', id: 'Disbursement Error' },
  { label: 'Other', id: 'Other' },
];
