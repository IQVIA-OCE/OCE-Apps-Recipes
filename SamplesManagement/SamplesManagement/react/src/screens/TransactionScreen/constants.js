export const packageConditions = [
  { label: 'Undamaged', id: '1' },
  { label: 'Damaged', id: '2' },
  { label: 'Opened', id: '3' },
];
