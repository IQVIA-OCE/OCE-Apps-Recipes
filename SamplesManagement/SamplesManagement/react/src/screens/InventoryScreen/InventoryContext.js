import React, { useContext, useEffect, useState } from 'react';
import {
  fetchActiveLotsProducts,
  fetchConfigs,
  fetchInventoryById,
  fetchInventoryDetail,
  fetchInventoryTypes,
  fetchLastSubmittedInventory,
  fetchTransactionDetails,
} from '../../api/Inventories';
import { normalizer } from '../../utils/utils';
import { AppContext } from '../../../AppContext';
import {
  getConfig,
  normalizeProductsList,
  normalizeProductsHistoryList,
} from './utils';
import { withNavigation } from 'react-navigation';
import { INVENTORY_STATUS } from './constants';
import { normalizeRecordTypes } from '../DashboardScreen/InventoriesWidget/utils';
import moment from 'moment';
import { INVENTORY_FORM_TYPE } from '../../constants/Inventories';

const initialValue = {
  editingType: INVENTORY_FORM_TYPE.preview,
  isLoading: true,
  preselectedProducts: [],
  products: [],
  productsHistory: [],
  error: null,
  config: {},
  lastInventoryCreatedDate: null,
  transactionDetails: [],
  recordType: {},
  formInitialValues: {},
};
export const InventoryContext = React.createContext(initialValue);

const inventoryKeys = {
  CreatedDate: 'createdDate',
  Id: 'id',
  OCE__Status__c: 'status',
  RecordTypeId: 'recordTypeId',
  OCE__Comments__c: 'comments',
  Name: 'name',
  OCE__Reason__c: 'reason',
  OCE__Auditor__c: 'auditor',
};

const configKeys = {
  OCE__SIShowSystemCalculatedFields__c: 'showCalculatedFields',
  OCE__SIShowSystemCount__c: 'showSystemCount',
  SetupOwnerId: 'ownerId',
  OCE__SIShowLoggedInUserRecords__c: 'record',
  OCE__STFinalStatus__c: 'status',
  OCE__SIHistoryListHidden__c: 'historyHidden',
};

const ContextProvider = ({ children, navigation }) => {
  const [state, setState] = useState(initialValue);
  const { userId, profileId, orgId } = useContext(AppContext);

  const handleFetchConfig = async () => {
    const [configsData] = await fetchConfigs(userId, profileId, orgId);

    const configs = normalizer(configKeys)(configsData);
    return getConfig(configs, [userId, profileId, orgId]);
  };

  const handleEditingType = editingType =>
    setState(prevState => ({ ...prevState, editingType }));

  useEffect(() => {
    handlePrepareData();
  }, []);

  const handlePrepareData = async () => {
    try {
      let inventory,
        transactionDetails,
        lastSubmittedInventoriesDetails,
        inventoryDetails,
        products,
        preselectedProducts,
        lastInventoryCreatedDate;
      const config = await handleFetchConfig();

      const [recordTypesData] = await fetchInventoryTypes();
      const recordTypes = normalizeRecordTypes(recordTypesData);

      const [lastInventories] = await fetchLastSubmittedInventory(
        userId,
        config.record
      );

      if (lastInventories.length) {
        inventory = [lastInventories[lastInventories.length - 1]];

        lastInventoryCreatedDate = inventory[0].CreatedDate;
        [lastSubmittedInventoriesDetails] = await fetchInventoryDetail(
          inventory[0].Id,
          true
        );
      } else {
        inventory = [
          {
            status: INVENTORY_STATUS.inProgress,
          },
        ];
      }

      if (navigation.state.params.id) {
        // preview or edit existing inventory
        [inventory] = await fetchInventoryById(navigation.state.params.id);
        [inventoryDetails] = await fetchInventoryDetail(
          navigation.state.params.id,
          state.editingType === INVENTORY_FORM_TYPE.preview
            ? false
            : inventory[0].OCE__Status__c !== INVENTORY_STATUS.inProgress
        );
      }

      [inventory] = normalizer(inventoryKeys)(inventory);

      if (inventory.status === INVENTORY_STATUS.inProgress) {
        [transactionDetails] = await fetchTransactionDetails(
          null,
          userId,
          config.status || 'Submitted'
        );
      } else {
        [transactionDetails] = await fetchTransactionDetails(
          navigation.state.params.id
        );
      }
      let productsHistory = normalizeProductsHistoryList(transactionDetails);

      const [productsList] = await fetchActiveLotsProducts();

      [products, preselectedProducts] = normalizeProductsList(
        productsList,
        lastSubmittedInventoriesDetails,
        inventoryDetails,
        transactionDetails,
        navigation.state.params.id ? inventory.status : null,
        Boolean(navigation.state.params.id)
      );

      setState(prevState => {
        let state = {
          ...prevState,
          isLoading: false,
          products,
          productsHistory,
          config,
          transactionDetails,
          lastInventoryCreatedDate,
        };
        if (navigation.state.params.id) {
          const recordType = recordTypes.all.find(
            el => el.Id === inventory.recordTypeId
          );

          state.editingType =
            navigation.state.params.type || INVENTORY_FORM_TYPE.preview;
          state.recordType = recordType;
          state.formInitialValues = {
            ...inventory,
            datetime: inventory.createdDate,
            products: preselectedProducts,
            buttonPressed: '',
          };
        } else {
          state.editingType = INVENTORY_FORM_TYPE.edit;
          state.recordType = navigation.state.params.recordType;
          state.formInitialValues = {
            id: '',
            datetime: moment(),
            products: preselectedProducts,
            status: INVENTORY_STATUS.inProgress,
            buttonPressed: '',
          };
        }

        return state;
      });
    } catch (error) {
      console.warn({ error });
      setState(prevState => ({
        ...prevState,
        isLoading: false,
        error,
      }));
    }
  };

  return (
    <InventoryContext.Provider
      value={{
        ...state,
        actions: { handleEditingType },
      }}
    >
      {children}
    </InventoryContext.Provider>
  );
};

export default withNavigation(ContextProvider);
