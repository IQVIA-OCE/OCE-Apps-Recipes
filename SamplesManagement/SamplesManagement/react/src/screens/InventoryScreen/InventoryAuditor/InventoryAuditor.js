import React, { useContext, useEffect, useRef, useState } from 'react';
import { StyleSheet, Text, View } from 'react-native';

import { Search, Menu, red, Paragraph } from 'apollo-react-native';
import { useBoolean } from '../../../hooks';
import Loader from '../../../components/Loader/Loader';
import { fetchAuditorById, fetchAuditors } from '../../../api/Inventories';
import { useFormikContext } from 'formik';
import { INVENTORY_FORM_TYPE } from '../../../constants/Inventories';

import { InventoryContext } from '../InventoryContext';

const InventoryAuditor = () => {
  const { editingType } = useContext(InventoryContext);
  const {
    values,
    setFieldValue,
    touched,
    errors,
    setFieldTouched,
  } = useFormikContext();

  const timer = useRef(0);
  const [value, setValue] = useState('');
  const [isVisible, visibleActions] = useBoolean(false);
  const [list, setList] = useState([]);
  const [isLoading, loadingActions] = useBoolean(false);

  const fetchList = text => {
    setValue(text);
    try {
      if (timer.current) {
        clearTimeout(timer.current);
      }
      let auditors;
      if (text.trim() && text !== value) {
        visibleActions.setTrue();
        timer.current = setTimeout(async () => {
          loadingActions.setTrue();
          [auditors] = await fetchAuditors(text);
          setList(auditors);
          loadingActions.setFalse();
        }, 400);
      }
    } catch (e) {
      console.warn({ e });
    }
  };

  const setAuditor = async id => {
    try {
      const [auditor] = await fetchAuditorById(id);

      setValue(auditor[0].Name);
    } catch (e) {
      console.warn(e);
    }
  };

  useEffect(() => {
    if (values.auditor && !value) {
      setAuditor(values.auditor);
    }
  }, []);

  if (editingType !== INVENTORY_FORM_TYPE.edit) {
    return <Paragraph style={{ fontSize: 16 }}>{value}</Paragraph>;
  }

  return (
    <>
      <Menu
        visible={isVisible}
        onDismiss={visibleActions.setFalse}
        anchor={
          <View>
            <Search
              placeholder="Search"
              value={value}
              onChangeText={fetchList}
              style={[styles.search, errors.auditor && styles.error]}
              onIconPress={() => {
                setValue('');
                setFieldValue('auditor', undefined);
              }}
              helperText={
                (touched.auditor || touched.buttonPressed) && errors.auditor
                  ? errors.auditor
                  : null
              }
              error={
                (touched.auditor || touched.buttonPressed) && errors.auditor
                  ? errors.auditor
                  : null
              }
            />
            {isLoading && <Loader />}
          </View>
        }
      >
        {list.map((el) => (
          <Menu.Item
            key={el.Id}
            onPress={() => {
              setValue(el.Name);
              visibleActions.setFalse();
              setFieldValue('auditor', el.Id);
              setFieldTouched('auditor', true);
            }}
          >
            <Text>{el.Name}</Text>
          </Menu.Item>
        ))}
      </Menu>
    </>
  );
};

const styles = StyleSheet.create({
  search: {
    width: 180,
  },
  error: {
    borderColor: red,
  },
});

export default InventoryAuditor;
