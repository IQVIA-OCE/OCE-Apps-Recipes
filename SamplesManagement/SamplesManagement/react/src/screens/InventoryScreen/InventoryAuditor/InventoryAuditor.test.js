import React from 'react';
import { Formik } from 'formik';
import InventoryAuditor from './InventoryAuditor';
import { Provider, Search } from '@oce-apps/apollo-react-native';
import { InventoryContext } from '../InventoryContext';
import { INVENTORY_FORM_TYPE } from '../../../constants/Inventories';
import { render, act, fireEvent } from '@testing-library/react-native';

const Component = ({ editingType, auditor }) => (
  <Provider>
    <InventoryContext.Provider
      value={{
        editingType,
      }}
    >
      <Formik
        initialValues={{
          auditor,
          id: '123',
          status: 'Saved',
          products: [{ id: '1' }],
        }}
      >
        <InventoryAuditor />
      </Formik>
    </InventoryContext.Provider>
  </Provider>
);

describe('InventoryAuditor', () => {
  it('should render properly', () => {
    const { UNSAFE_root, getByText } = render(
      <Component editingType={INVENTORY_FORM_TYPE.edit} />
    );

    fireEvent(UNSAFE_root.findByType(Search), 'onIconPress');
    fireEvent.changeText(UNSAFE_root.findByType(Search), 'some text');
    act(() => jest.runAllTimers());

    expect(getByText(/District Manager/)).toBeTruthy();
  });

  it('should render readonly component', () => {
    const { getByText } = render(<Component editingType={INVENTORY_FORM_TYPE.editSaved} auditor="0056F00000A4qgKQAR" />);
    act(() => jest.runAllTimers());

    expect(getByText(/District Manager/)).toBeTruthy();
  });
});
