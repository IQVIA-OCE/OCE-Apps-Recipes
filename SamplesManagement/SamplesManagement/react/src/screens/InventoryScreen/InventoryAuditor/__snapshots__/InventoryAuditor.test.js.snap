// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryAuditor should render properly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "width": 176,
            },
            Array [
              Object {
                "width": "100%",
              },
              undefined,
              Array [
                Object {
                  "width": 180,
                },
                undefined,
              ],
            ],
            Object {
              "opacity": 1,
            },
            Object {
              "elevation": 10,
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 5,
                "width": 0,
              },
              "shadowOpacity": 0.34,
              "shadowRadius": 6.27,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "borderColor": "#D9D9D9",
                "borderRadius": 4,
                "borderWidth": 1,
              },
              Object {
                "width": 180,
              },
            ]
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": 40,
                "minHeight": 40,
                "overflow": "hidden",
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          >
            <View
              accessible={true}
              focusable={false}
              isTVSelectable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "left": 10,
                  "opacity": 1,
                  "position": "absolute",
                  "right": "auto",
                  "top": "50%",
                  "transform": Array [
                    Object {
                      "translateY": -12.5,
                    },
                  ],
                  "zIndex": 5,
                }
              }
            >
              <Text
                accessibilityElementsHidden={true}
                allowFontScaling={false}
                importantForAccessibility="no-hide-descendants"
                pointerEvents="none"
                style={
                  Array [
                    Object {
                      "color": "#999",
                      "fontSize": 25,
                    },
                    Array [
                      Object {
                        "transform": Array [
                          Object {
                            "scaleX": 1,
                          },
                        ],
                      },
                      Object {
                        "backgroundColor": "transparent",
                      },
                    ],
                    Object {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                Ôçâ
              </Text>
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              maxResizableWidth={750}
              minHeight={60}
              minWidth={0}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              paddingLeft={40}
              paddingRight={8}
              placeholder="Search"
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              rejectResponderTermination={true}
              selectionColor="#0768FD"
              style={
                Array [
                  Object {
                    "margin": 0,
                    "textAlign": "left",
                    "zIndex": 5,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "height": 40,
                    "minHeight": 0,
                    "paddingLeft": 40,
                    "paddingRight": 8,
                    "paddingVertical": 5,
                    "textAlignVertical": "center",
                    "width": "100%",
                  },
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignContent": "center",
            "backgroundColor": "rgba(255,255,255,0.7)",
            "flexDirection": "row",
            "height": "100%",
            "justifyContent": "center",
            "position": "absolute",
            "width": "100%",
            "zIndex": 100,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Object {
                "height": 24,
                "opacity": 1,
                "width": 24,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "height": 24,
                    "transform": Array [
                      Object {
                        "rotate": "765deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 12,
                      "overflow": "hidden",
                      "width": 24,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "height": 24,
                        "transform": Array [
                          Object {
                            "translateY": 0,
                          },
                          Object {
                            "rotate": "-165deg",
                          },
                        ],
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "height": 12,
                          "overflow": "hidden",
                          "width": 24,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "borderColor": "#015ff1",
                            "borderRadius": 12,
                            "borderWidth": 2.4,
                            "height": 24,
                            "width": 24,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "bottom": 0,
                  "justifyContent": "center",
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <View
                style={
                  Object {
                    "height": 24,
                    "transform": Array [
                      Object {
                        "rotate": "765deg",
                      },
                    ],
                    "width": 24,
                  }
                }
              >
                <View
                  collapsable={false}
                  style={
                    Object {
                      "height": 12,
                      "overflow": "hidden",
                      "top": 12,
                      "width": 24,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "height": 24,
                        "transform": Array [
                          Object {
                            "translateY": -12,
                          },
                          Object {
                            "rotate": "345deg",
                          },
                        ],
                        "width": 24,
                      }
                    }
                  >
                    <View
                      collapsable={false}
                      style={
                        Object {
                          "height": 12,
                          "overflow": "hidden",
                          "width": 24,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "borderColor": "#015ff1",
                            "borderRadius": 12,
                            "borderWidth": 2.4,
                            "height": 24,
                            "width": 24,
                          }
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`InventoryAuditor should render readonly component 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
          "writingDirection": "ltr",
        },
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 14,
              "letterSpacing": 0.25,
              "lineHeight": 20,
              "marginVertical": 2,
            },
            Object {
              "fontSize": 16,
            },
          ],
          Array [
            Object {
              "fontSize": 14,
              "letterSpacing": 0.25,
              "lineHeight": 20,
              "marginVertical": 2,
            },
            Object {
              "fontSize": 16,
            },
          ],
        ],
      ]
    }
  >
    
  </Text>
</View>
`;
